### DISCUSSION TITLE
방만들기 CRUD 백엔드 기술 스택 조사

### DISCUSSION THREAD

**Discussion by Junroot** [Junroot/Nomat#30] [개발]
방만들기 CRUD 백엔드 기술 스택 조사

**Comment by Junroot** [comment]
## REST API: Spring Boot 3.2.1
  - https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter
  - Spring Boot 버전 중 stable한 최신 버전
  - 국내에서 가장 흔하게 사용하는 웹 프레임워크
  - 학습 목적으로 선택

**Reply by Junroot** [reply]
- 사용 언어: Kotlin

**Comment by Junroot** [comment]
## WebSocket

- Spring WebSocket, Spring Messaging
- WebSocket 통신 추상화 제공, 메시지 브로커 추상화 제공, WebSocket 지원하지 않는 경우에대한 polling 방식 제공(Sock.js)
- https://mvnrepository.com/artifact/org.springframework/spring-websocket/6.1.3
- https://mvnrepository.com/artifact/org.springframework/spring-messaging/6.1.3
- https://www.baeldung.com/websockets-spring

**Comment by Junroot** [comment]
## DBMS 및 ORM

- MariaDB 또는 MySQL 
- Spring data JPA 3.2.1
- https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-data-jpa/3.2.1

**Comment by Junroot** [comment]
## 로깅 관리

- Loki
- 가장 많이 사용하는 로깅 관리 시스템으로는 ELK와 Loki가 있음
- Loki는 모든 줄에 대한 인덱싱을 하지 않고, 로깅과 관련된 라인만 인덱싱한다.

**Comment by Junroot** [comment]
## 컨테이너 배포

- ECS? EKS? 조사 필요

**Reply by hsrst** [reply]
### EKS (Elastic Kubernetes Service)
Kubernetes를 통해 컨테이너 오케스트레이션을 수행함.

### ECS (Elastic Container Service)
AWS의 자체적인 소프트웨어를 통해 컨테이너 오케스트레이션을 수행함.

**Comment by hsrst** [comment]
## 데이터 플레인
### EC2
컨테이너의 자원을 이용자가 직접 관리할 수 있음.

장점
- 가상 머신 방식으로, 스펙 (CPU 코어 수, 메모리, 스토리지 등)을 비교적 자유롭게 선택 가능
- OS에 직접 개입 가능
- 컨테이너 이미지를 EC2 내에 캐싱할 수 있어서 Fargate 타입에 비해 배포가 빠름

단점
- 호스트를 관리 (스케일링이나 패치 등)할 필요가 있음

### Fargate
서버리스 방식의 데이터 플레인으로, 컨테이너의 자원을 AWS 측에서 관리해 줌.

장점
- 호스트를 관리할 필요가 없음

단점
- EC2에 비해 스펙 설정에 제약이 있음
- OS에 직접 개입할 수 없음
- 컨테이너 이미지가 캐싱되지 않으므로 EC2에 비해 배포가 느림
- EC2에 비해 가격이 비쌈

출처: https://zenn.dev/issy/articles/zenn-ecs-vs-eks

**Comment by Junroot** [comment]
## EC2 인스턴스

- 개발 환경: 최대한 프리티어 이용
- 운영 환경: t3.medium(차후에 트래픽 예측해서 결정)

**Comment by Junroot** [comment]
- 쿠버네티스를 사용할 것인가?: YES
  - 컨테이너 관리 용이
  - 쿠버네티스 자체에서 제공해주는 배포 전략, 모니터링 기능
- EKS, ECS는 사용하지 않는다.
  - 너무 비싼 가격
- EC2 인스턴스에 쿠버네티스 직접 구축하는 방식으로 결정

**Comment by Junroot** [comment]
## DB 인프라 구성

- 옵션1: RDS
- 옵션2: EC2에 DB 직접 설치

RDS가 EC2보다 운영 관리 면에서는 편리하지만, 비용이 대략적으로 1.5배 정도 더 나올것으로 추측.
비용 절감을 위해 옵션2 방식으로 운영하기로 결정

참고자료: https://www.smileshark.kr/post/amazon-ec2-vs-amazon-rds-how-to-choose-right-hosted-database#viewer-netf

**Comment by Junroot** [comment]
개발: WAS(프리티어 EC2), DB(t3.medium)
운영: WAS(t3.medium), DB(t3.medium)
